### PR TITLE
Adds Async support

### PR DESCRIPTION
**Pull request by bradlangel**

### CODE REVIEW

File: statsd_client.ml [hunk 1]
```diff
@@ -153,3 +152,47 @@ module Lwt = Make (
...
+        | Unix.PF_INET, Unix.SOCK_DGRAM, 17 -> Socket.Type.udp
+        | Unix.PF_INET, Unix.SOCK_STREAM, 6 -> Socket.Type.tcp
+        | _ -> failwith "Protocol not recognized"
+      in
+      Socket.fd (Socket.create socket_type)
+
+    let inet_addr_exn = function
+      | Unix.ADDR_INET (inet_addr, port) ->
+        Socket.Address.Inet.create inet_addr ~port
+      | Unix.ADDR_UNIX _ -> failwith "Inet addr does not exist"
```
Comment (kevinqiu) [inline statsd_client.ml]: This message seems a bit cryptic. What about something like "Only Internet addresses supported" or "Unix domain addresses not supported".

File: statsd_client.ml [hunk 2]
```diff
@@ -20,7 +20,6 @@ module Make(U : sig
   (** include Lwt_unix or Unix depending on your use *)
```
Comment (kevinqiu) [inline statsd_client.ml]: Should we update these functor comments with Async references? Or remove them entirely?

**Review by kevinqiu — APPROVED** [review]
2 optional comment changes, code looks good to me! Yay functors! Sad that we need to because of Async + Lwt tho :(

File: statsd_client.ml [hunk 3]
```diff
@@ -153,3 +152,47 @@ module Lwt = Make (
...
+    let list_iter f l = Deferred.List.iter ~f l
+    let close fd = Unix.close fd
+
+    type file_descr = Async_unix.Fd.t
+
+    let socket domain typ protocol =
+      let socket_type =
+        match domain, typ, protocol with
+        | Unix.PF_INET, Unix.SOCK_DGRAM, 17 -> Socket.Type.udp
+        | Unix.PF_INET, Unix.SOCK_STREAM, 6 -> Socket.Type.tcp
```
Comment (brendanlong) [inline statsd_client.ml]: What are 17 and 6?

File: statsd_client.ml [hunk 4]
```diff
@@ -153,3 +152,47 @@ module Lwt = Make (
...
+        Socket.Address.Inet.create inet_addr ~port
+      | Unix.ADDR_UNIX _ -> failwith "Inet addr does not exist"
+
+    let sendto fd msg offset msg_length flags socket_address =
+      match Udp.sendto () with
+      | Error e -> return (-1)
+      | Ok send ->
+        let buf = msg |> Iobuf.of_string |> Iobuf.read_only in
+        let inet_addr = inet_addr_exn socket_address in
+        send fd buf inet_addr >>| fun () -> msg_length
```
Comment (brendanlong) [inline statsd_client.ml]: Not a fan of multi-statement lines like this. It's much easier to read them like this:

```
send fd buf inet_addr
>>| fun () ->
msg_length
```

Or maybe in this case:

```
send fd buf inet_addr
>>| fun () -> msg_length
```

**Review by brendanlong — APPROVED** [review]
Make sure to open a pull request upstream too.

File: statsd_client.ml (hunk 2 above)
Comment (bradlangel) [reply statsd_client.ml]: I removed the comment.

File: statsd_client.ml (hunk 3 above)
Comment (bradlangel) [reply statsd_client.ml]: IP protocol numbers. https://en.wikipedia.org/wiki/List_of_IP_protocol_numbers
Put a comment and renamed `protocol` to `ip_protocol_num`

**Pull request merged**